### PR TITLE
Add `keys` functionality (for key generation, saving, and signing).

### PR DESCRIPTION
**Pull request by zanebeckwith**
This PR implements a new `keys` functionality, which includes:
- Child key generation
- Checking if a parent handle is usable (and if not creating a parent at that handle)
- Loading a key into the TPM
- Signing using a key
- Getting the raw public key of a TPM key
- Writing a TPM key as a PEM-encoded file

In the grand tradition of this repo, this functionality is only tested on a few happy paths, and low-level byte-wrangling code has no other tests... However, I have used a key generated (and written) using this code in James Bottomley's OpenSSL engine, to create a self-signed certificate (for some reason, I'm having a hard time getting the tpm2-software engine to use an ECDSA key, at least on my VM. But, this has nothing to do with `xaptum-tpm`)

Passwords are not set on any of the created keys. However, operations that require *hierarchy* authorization (creation of a new primary key, persistence of a key to NVRAM) allow a password to be specified, in case the user has set a password on the OWNER hierarchy (in which all the keys are created, by default).

By default, keys are created in the OWNER hierarchy, and the parent is located at `0x800001`. James Bottomley claims that many TPM manufacturers ship their chips with a parent handle already stored at that index, in  which case we're able to just use that key transparently.

Only ECDSA (secp256r1) keys are created, and there are a couple of places in the code where this is assumed. 

The intended use of this new API, in the `enftun` (and `enftun-keygen`), is:
- Use `xtpm_gen_key` and `xtpm_write_key` in `enftun-keygen` to generate a new key
  - The resulting file can then be given to the OpenSSL code for creating the self-signed certificate
  - If the user has set a password on the OWNER hierarchy, they can specify it here (manually, presumably)
  - This results in a parent key that's persisted to the TPM's NVRAM, and a child key that's output to file as a PEM-encoded encrypted blob
- When running the TLS handshake in `enftun`:
  - Read in the key file
  - Use the OpenSSL library to parse the PEM and ASN.1, in order to get the `parent` field from the key file
    - I didn't include any analogous functionality in this code, because writing an ASN.1 parser is a much larger/hairier task, and in this one usecase we already know we have access to OpenSSL
  - Use the `xtpm_create_parent` function to ensure the parent is available in the TPM
    - If the user has set a password on the OWNER hierarchy, and the parent key is *not* available, then the re-creation of the parent key will fail
    - This code does provide the option that they could specify the password in a config file provided to `enftun`, but that's obviously a dumb idea. So, I don't really know if there's any good way to handle that scenario
    - But, just to be clear, the happy path (the parent *does* exist and is OK), doesn't require a password
  - Pass the key file to the OpenSSL library as usual

I'm working now on updating `xtt` to use this functionality.

I wanted to see what you guys think of the API

### CODE REVIEW

**Review by drbild — COMMENTED** [review]
API looks good to me.

How is the xtpm_parent_conf struct initialized?  Maybe I just missed that function in the public API, looking at this on my phone.

**Comment by dberliner** [comment]
This looks like a very good interface. Converting these structs into OpenSSL objects will be another task and it looks like Bottomleys engine has a relatively standalone function to accomplish this.

**Comment by zanebeckwith** [comment]
> How is the xtpm_parent_conf struct initialized?

Manually.

Of the three members of the struct, two are integer handles (so easily set via simple assignment) and the other is a sized buffer for the password.

Honestly, that struct was a crutch I used during development, as I figured out what the function signatures needed to be, then *ex post facto* I decided it was a nice consolidation. But, looking at it now, it might be better to explode it and take the parameters individually (and make the password a `const char*` and `size_t`)

I think I'm going to make that change now

**Comment by zanebeckwith** [comment]
> Converting these structs into OpenSSL objects will be another task

My thought was that we wouldn't need to do that. Instead, we can just specify the PEM key file name to the OpenSSL code, and let OpenSSL do its thing loading it

During the TLS handshake, after using `xtpm_create_parent` to ensure the parent is available, giving the file name to OpenSSL is all `enftun` would need to do (the `enftun` config would just have the key file name, and some flag indicating to use the TPM).

During `keygen`, we would run through the steps to generate the key and write it to file, then just use the same existing code for certificate generation but 1) load the TPM OpenSSL engine and 2) specify the newly-created key file as the name of the key

In the `keygen` case, that results in a little bit of unnecessary processing repetition (we'll have the key information in memory, in the `xtpm_key` struct, but then force OpenSSL to still load the key from file), but that's not a fast path operation, so avoiding the extra development effort for us seems like a good idea to me

But I might be missing something or not thinking this through

**Comment by zanebeckwith** [comment]
I've updated this branch, getting rid of the `parent_conf` struct. I actually do like this much better. 

I also fixed the CI. `cppcheck` was complaining that a function wasn't used (because I had no test whatsoever of the `get_public_key` function. So, I added one. `cppcheck` has been keeping me honest about not testing my functions recently. Someone should give a presentation about static analysis tools...

**Comment by zanebeckwith** [comment]
Update 2: I just pushed two changes to this:
- Use `const` appropriately in the new keys functions (previously, I forgot to annotate any of the params as const
  - I ran into this while integrating with XTT, where the compiler started complaining, reasonably
- Add a `xtpm_flush_key` function, and run `Tss2_Sys_Flush` after:
  - 1) creating the parent key (when we EvictControl, so the key is persisted, the temporary copy is still around, taking up memory)
  - 2) signing (we load the key to sign, so after signing we need to flush it)

Note: Because the new flush functionality requires the changes from teh `zanebeckwith/fixes` branch, which is still under review, the CI will fail at the moment (until we merge the `fixes` PR, and rebase this branch onto it)

File: include/xaptum-tpm/keys.h [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+ *  - parent_handle = 0x81000001 (set to 0 to use default)
+ *  - hierarchy = TPM2_RH_OWNER (set to 0 to use default)
+ *  - hierarchy_password = empty auth
+ *  - hierarchy_password_length = 0 (set to 0 to use default password)
+ *
+ * Returns `TSS2_RC_SUCCESS` if parent is OK (either already loaded, or we created it),
+ * other if not (e.g. existing key is bad, or we're unable to create a new one).
+ */
+TSS2_RC
+xtpm_create_parent(TSS2_TCTI_CONTEXT *tcti_ctx,
```
Comment (drbild) [inline include/xaptum-tpm/keys.h:62]: Maybe `xtpm_ensure_parent`?

**Review by drbild — COMMENTED** [review]
> * Read in the key file
> * Use the OpenSSL library to parse the PEM and ASN.1, in order to get the `parent` field from the key file
>   
>   * I didn't include any analogous functionality in this code, because writing an ASN.1 parser is a much larger/hairier task, and in this one usecase we already know we have access to OpenSSL
> * Use the `xtpm_create_parent` function to ensure the parent is available in the TPM
>   
>   * If the user has set a password on the OWNER hierarchy, and the parent key is _not_ available, then the re-creation of the parent key will fail
>   * This code does provide the option that they could specify the password in a config file provided to `enftun`, but that's obviously a dumb idea. So, I don't really know if there's any good way to handle that scenario
>   * But, just to be clear, the happy path (the parent _does_ exist and is OK), doesn't require a password

Doesn't the TPM engine take care of [loading the parent]( https://github.com/tpm2-software/tpm2-tss-engine/blob/47eb165b4a4f13f51d26d67687de62281be8891f/src/tpm2-tss-engine-common.c#L563) key? So `enftun` wouldn't need to do that.    It won't handle the case where the parent handle needs a password, but I agree that is dumb idea for this use case anyway.

Otherwise, the code and API LGTM.

**Comment by zanebeckwith** [comment]
> Doesn't the TPM engine take care of [loading the parent](https://github.com/tpm2-software/tpm2-tss-engine/blob/47eb165b4a4f13f51d26d67687de62281be8891f/src/tpm2-tss-engine-common.c#L563) key

The `Esys_TR_FromTPMPublic` function "loads" a key that's already available at a persistent handle. My intention with having `enftun` always call `xtpm_create_parent` is to cover the corner case where the parent has been deleted, so we need to recreate it and re-persist it to that handle, i.e. to make sure the subsequent call to `Esys_TR_FromTPMPublic` succeeds

I admit that's an odd scenario: that handle has been deleted, but it hasn't been replaced with something else (so we can successfully save our parent key there again) nor has the hierarchy seed been regenerated (i.e. `Clear` hasn't been called)

In the typical case, where the parent key is still available at that handle, the `xtpm_create_parent` just calls `Tss2_Sys_ReadPublic`, so the overhead is one extra round-trip with the TPM

If we decide to not use `xtpm_create_parent` in `enftun`, though, I'll remove it from this project, since it's not used anywhere else and otherwise confuses the API

**Comment by drbild** [comment]
> My intention with having `enftun` always call `xtpm_create_parent` is to cover the corner case where the parent has been deleted, so we need to recreate it and re-persist it to that handle, i.e. to make sure the subsequent call to `Esys_TR_FromTPMPublic` succeeds

`enftun` recreating a deleted parent when trying to connect feels wrong (or at least, unnecessarily complex) to me.  Feel free to push back thought.

I think it does make sense to create the parent when creating the child key.  So the two cases:

### 1. Manual creation with `enftun-keygen`

The child key is created manually by running `enftun-keygen`.  This process can (re)create the parent, if needed.

If the parent is later deleted, someone will have to rerun `enftun-keygen`.  This seems fair.  You manually provisioned the key; if you mess with the TPM, you'll have to manually re-provision the key.

### 2. Automated creation with `xtt`

The child key is created automatically by `xtt` (triggered by `enftun`).  `xtt` will (try to) (re)create the parent, if needed.  The `enftun` config just needs to expose the right knobs to the user, as you said, to pass to `xtt`. I think just the hierarchy and parent handle is sufficient for now (with most users leaving them at the defaults).

If the parent handle is deleted, then the TLS connection will fail and `enftun` will retry the `xtt` handshake. This will implicitly recreate the parent handle, if possible. So we get the same affect, without `enftun` having to parse the cert, extract the parent handle, etc.

Thoughts?

**Comment by zanebeckwith** [comment]
Actually, yes, that makes sense. I've still been in the mindset of the DAA key distribution on the TPMs, where loss of the key is a complete denial of service. I wasn't thinking about how enftun will just kick back to the XTT handshake, which will regenerate the key and its parent if need-be

So, yes, I agree with you.

I've removed the `xtpm_create_parent`. There were multiple things about how to expose that functionality that had already bugged me (as evidenced by its weird name), so I'm happy to remove it

**Comment by drbild** [comment]
> I've removed the `xtpm_create_parent`. There were multiple things about how to expose that functionality that had already bugged me (as evidenced by its weird name), so I'm happy to remove it

Do we still need an equivalent function for `xtt` and `enftun-keygen` to use?

**Review by drbild — COMMENTED** [review]
Do we still need an equivalent function for `xtt` and `enftun-keygen` to use?

**Comment by zanebeckwith** [comment]
> Do we still need an equivalent function for `xtt` and `enftun-keygen` to use?

No. The `xtpm_gen_key` already checks whether the given `parent` exists (and can be used as a parent key), and creates it if not.

That `xtpm_create_parent` was *only* for the purpose of "I have an already-created key, make sure the parent it claims is actually usable to decrypt it". 

tl;dr Both `xtt` and `enftun-keygen` can just use `xtpm_gen_key` to create the key, and that will ensure the parent is persisted at the correct handle

**Review by drbild — APPROVED** [review]
Got it.  LGTM.
